File: MiniDB_Projects/Team_candyman/internal/sql/ast.go
// Package sql implements MiniDB's SQL front end: a hand-written lexer and a
// recursive-descent parser producing a small AST. The supported grammar is
// deliberately compact (CREATE TABLE, INSERT, DELETE, SELECT with WHERE/JOIN/
// GROUP BY, and BEGIN/COMMIT/ROLLBACK) so every production is explainable.
package sql

import "minidb/internal/types"

// Statement is any top-level SQL statement.
type Statement interface{ stmt() }

// ColumnDef is a column declaration in CREATE TABLE.
type ColumnDef struct {
	Name       string
	Type       types.Type
	PrimaryKey bool
}

// CreateTable is `CREATE TABLE name (cols...)`.
type CreateTable struct {
	Table   string
	Columns []ColumnDef
}

// Insert is `INSERT INTO name [(cols)] VALUES (...) [, (...)]`.
type Insert struct {
	Table   string
	Columns []string // optional explicit column list; empty means all, in order
	Rows    [][]Expr
}

// Delete is `DELETE FROM name [WHERE expr]`.
type Delete struct {
	Table string
	Where Expr
}

// TableRef names a table with an optional alias.
type TableRef struct {
	Name  string
	Alias string // defaults to Name
}

// SelectItem is one projected expression with an optional output name.
type SelectItem struct {
	Expr  Expr
	Alias string
}

// Select is the query statement. Star is true for `SELECT *`. All join and where
// predicates are collected into Predicates (ANDed); the planner decides join order.
type Select struct {
	Star       bool
	Items      []SelectItem
	Tables     []TableRef
	Predicates []Expr
	GroupBy    []Expr
	Explain    bool // EXPLAIN prefix: print the chosen plan instead of running it
}

// Begin / Commit / Rollback control explicit transactions.
type Begin struct{}
type Commit struct{}
type Rollback struct{}

func (*CreateTable) stmt() {}
func (*Insert) stmt()      {}
func (*Delete) stmt()      {}
func (*Select) stmt()      {}
func (*Begin) stmt()       {}
func (*Commit) stmt()      {}
func (*Rollback) stmt()    {}

// Expr is an expression node.
type Expr interface{ expr() }

// ColumnRef references a column, optionally qualified by a table/alias.
type ColumnRef struct {
	Table string // optional qualifier
	Name  string
}

// Literal is a constant value.
type Literal struct{ Value types.Value }

// BinaryExpr covers comparisons (=, !=, <, <=, >, >=) and the AND connective.
type BinaryExpr struct {
	Op    string
	Left  Expr
	Right Expr
}

// AggCall is an aggregate function call, e.g. COUNT(*), SUM(col).
type AggCall struct {
	Func string // COUNT, SUM, AVG, MIN, MAX
	Star bool    // COUNT(*)
	Arg  Expr    // argument when not star
}

func (*ColumnRef) expr()  {}
func (*Literal) expr()    {}
func (*BinaryExpr) expr() {}
func (*AggCall) expr()    {}

File: MiniDB_Projects/Team_QuorumDB/src/minidb/sql/ast.py
"""Abstract syntax tree for MiniDB's SQL subset.

The parser turns SQL text into these dataclasses, which the planner then turns
into an execution plan. The supported surface is intentionally small but
complete enough to demonstrate end-to-end query processing:

    CREATE TABLE / CREATE INDEX / DROP TABLE
    INSERT INTO ... VALUES ...
    DELETE FROM ... [WHERE ...]
    SELECT ... FROM ... [JOIN ... ON ...] [WHERE ...]
    BEGIN / COMMIT / ROLLBACK

Predicates are conjunctions (AND) of simple comparisons; each comparison has a
column on the left and either a literal or another column (for joins) on the
right.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, List, Optional


# -- expressions -----------------------------------------------------------
@dataclass(frozen=True)
class ColumnRef:
    name: str
    table: Optional[str] = None      # qualifier, e.g. the "u" in u.id

    @property
    def qualified(self) -> str:
        return f"{self.table}.{self.name}" if self.table else self.name


@dataclass(frozen=True)
class Literal:
    value: Any                       # int | float | str | bool | None


@dataclass(frozen=True)
class Comparison:
    left: ColumnRef
    op: str                          # one of = != < <= > >=
    right: Any                       # Literal or ColumnRef

    @property
    def is_join_predicate(self) -> bool:
        return isinstance(self.right, ColumnRef)


@dataclass
class Predicate:
    """A conjunction (AND) of comparisons. Empty means 'always true'."""
    comparisons: List[Comparison] = field(default_factory=list)

    def __bool__(self) -> bool:
        return bool(self.comparisons)


# -- statements ------------------------------------------------------------
@dataclass
class ColumnDef:
    name: str
    type: str
    not_null: bool = False
    primary_key: bool = False


@dataclass
class CreateTable:
    name: str
    columns: List[ColumnDef]
    pk_column: Optional[str] = None


@dataclass
class CreateIndex:
    table: str
    column: str
    unique: bool = False
    name: Optional[str] = None


@dataclass
class DropTable:
    name: str


@dataclass
class Insert:
    table: str
    columns: Optional[List[str]]     # None means "all columns, in order"
    rows: List[List[Any]]


@dataclass
class Delete:
    table: str
    where: Predicate = field(default_factory=Predicate)


@dataclass
class JoinClause:
    table: str
    on: Comparison
    alias: Optional[str] = None

    @property
    def qualifier(self) -> str:
        return self.alias or self.table


@dataclass
class Select:
    columns: List[str]               # ["*"] or explicit (possibly qualified)
    from_table: str
    from_alias: Optional[str] = None
    joins: List[JoinClause] = field(default_factory=list)
    where: Predicate = field(default_factory=Predicate)

    @property
    def from_qualifier(self) -> str:
        return self.from_alias or self.from_table


@dataclass
class Begin:
    pass


@dataclass
class Commit:
    pass


@dataclass
class Rollback:
    pass
